### ISSUE TITLE
Type checking error with array types and optional properties

### ISSUE THREAD

**Issue by henryptung** [Microsoft/TypeScript#4392]
Seems like there's something wrong with type-checking when it comes to arrays of an interface type with optional properties.  When the first element has the property with incorrect type, it catches the error:
```
$ cat wrong.ts
interface a { b?: number; }
let x: a[] = [{b: "x"}];
$ tsc wrong.ts
x.ts(2,5): error TS2322: Type '{ b: string; }[]' is not assignable to type 'a[]'.
  Type '{ b: string; }' is not assignable to type 'a'.
    Types of property 'b' are incompatible.
      Type 'string' is not assignable to type 'number'.
```
but if the first element omits it and the second has incorrect type, tsc happily ignores it:
```
$ cat omittedWrong.ts
interface a { b?: number; }
let x: a[] = [{}, {b: "x"}];
$ tsc omittedWrong.ts
$ cat omittedWrong.js
var x = [{}, { b: "x" }];
```

Few other cases: If the first element has the property and is correct, tsc catches the error:
```
$ cat correctWrong.ts
interface a { b?: number; }
let x: a[] = [{b: 1}, {b: "x"}];
$ tsc correctWrong.ts
w.ts(2,5): error TS2322: Type '({ b: number; } | { b: string; })[]' is not assignable to type 'a[]'.
  Type '{ b: number; } | { b: string; }' is not assignable to type 'a'.
    Type '{ b: string; }' is not assignable to type 'a'.
      Types of property 'b' are incompatible.
        Type 'string' is not assignable to type 'number'.
```
but it doesn't catch it as long as there's one entry before the incorrect one omitting the property:
```
$ cat mixedWrong.ts
interface a { b?: number; }
let x: a[] = [{b: 1}, {b: 2}, {}, {b: 3}, {b: "x"}];
$ tsc mixedWrong.ts
$ cat mixedWrong.js
var x = [{ b: 1 }, { b: 2 }, {}, { b: 3 }, { b: "x" }];

```

Hope that helps!

**Comment by henryptung** [comment]
Also:
```
$ tsc --version
message TS6029: Version 1.5.3
```

**Comment by RyanCavanaugh** [comment]
This has been fixed in `master`:
```ts
interface a { b?: number; }
let x: a[] = [{}, {b: "x"}];
```

> a.ts(2,5): error TS2322: Type '({} | { b: string; })[]' is not assignable to type 'a[]'.
  Type '{} | { b: string; }' is not assignable to type 'a'.
    Type '{ b: string; }' is not assignable to type 'a'.
      Types of property 'b' are incompatible.
        Type 'string' is not assignable to type 'number'.
